### PR TITLE
Hindsight Experience Replay

### PR DESCRIPTION
**Pull request by prabhatnagarajan**

### CODE REVIEW

File: examples/her/train_dqn_bit_flip.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    def step(self, action):
+        # Compute action outcome
+        bit_new = int(not self.observation["observation"][action])
+        new_obs = self.observation["observation"].copy()
+        new_obs[action] = bit_new
+        # Set new observation
+        dg = self.observation["desired_goal"]
+        self.observation["desired_goal"] = dg.copy()
+        self.observation["achieved_goal"] = new_obs
+        self.observation["observation"] = new_obs
```
Comment (justindujardin) [inline examples/her/train_dqn_bit_flip.py:60]: I had to create a new object here for training to work. The successes should start showing up in the logs within 1000 episodes.

```suggestion
        self.observation = {
            "desired_goal": dg.copy(),
            "achieved_goal": new_obs,
            "observation": new_obs,
        }
```

If I had to guess why this change is needed, it'd be because the `self.observation` object is being inserted into the replay buffer, and when we later change the object keys it updates for all the observations in the replay buffer because they share a source object? Maybe I'm wrong about the reason, but the new object is needed to train in my tests.

Comment (prabhatnagarajan) [reply examples/her/train_dqn_bit_flip.py:60]: Thanks! You're right, I think we either have to do this or return `self.observation.copy(), reward, done, {}` instead of `self.observation, reward, done, {}`

**Comment by muupan** [comment]
Can you verify performance matches the published one or openai/baselines on Fetch tasks as well?

**Comment by prabhatnagarajan** [comment]
@ummavi also mentioned this in person. The main point of this PR is to add the HER feature for users, not to reproduce the Hindsight Replay paper, which has a whole different set of hyperparameters. I did the bit flip example because it's easy to implement, fast to experiment on, and a good verification that the system works. Moreover, this HER implementation is adapted from the one we used here: https://arxiv.org/abs/2007.08082, so correctness shouldn't be an issue (plus there are tests).

**Comment by muupan** [comment]
A verification that the system just "works" is not enough. A verification that the system works at least as well as the published result is needed for a new feature to keep pfrl reliable.

**Comment by prabhatnagarajan** [comment]
I'm not sure what you mean by `just "works"'. I've shown that it works on the bit-flip domain used in the paper (though they don't provide code or hyperparameters for that). 

To come up with somewhat a ridiculous analogy, if a paper uses a function that computes the standard deviation, you don't need to reproduce the whole paper's results to be sure that standard deviation works correctly. In the case of Hindsight Experience Replay, there's a quite a lot of tricks and modifications that they use in addition to HER, which has a relatively simple implementation. In effect, those changes sit well outside of Hindsight experience replay. There wouldn't be changes to HER hyperparameters, for example, to achieve that level of reproducibility. In other words, the nature of HER is that being "correct" means that it also works (which I think is what you meant when you said "just works").

What would be helpful is:
- If you could point out concretely what you think could be wrong with the HER implementation
- What you find unconvincing about the results

But I'm unclear about generically stating that an entire system should work as well as a published paper in order to add a piece of that paper. Especially since for other features like ACER and Noisy networks (https://github.com/chainer/chainerrl/pull/169#issuecomment-344176330), we didn't reproduce the paper to determine sufficient performance. In these case, that it "works" was sufficient. I'd appreciate if we were consistent across the board about our requirements for feature additions, especially in scenarios where correctness and it working go hand in hand.

I see your point, and it's certainly true that more results are always more reassuring. But I don't think it's necessary here, especially given our precedents, and I think if you read the code you might change your mind. It seems to me that you're insinuating that for any and all features, regardless of what the code shows or how simple/complex it is, we have to reproduce the entire system to add the feature. 
Moreover, I should reiterate that this implementation was used in our own paper(unless you doubt the correctness of that?). And again, seems inconsistent with what we've done in the past, and as we've had no internal discussion regarding this, don't see why we should randomly introduce this requirement out of the blue.

**Comment by muupan** [comment]
Let's continue in the internal chat.

**Comment by prabhatnagarajan** [comment]
Current results
PickAndPlace
Push
Slide

File: examples/her/train_dqn_bit_flip.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        self.clear_statistics()
+
+    def compute_reward(self, achieved_goal, desired_goal, info):
+        return reward_fn(desired_goal, achieved_goal)
+
+    def _check_done(self):
+        success = (
+            self.observation["desired_goal"] == self.observation["achieved_goal"]
+        ).all()
+        return (self.steps >= self.n) or success, success
```
Comment (prabhatnagarajan) [inline examples/her/train_dqn_bit_flip.py:44]: I think I should use the wrapper `ContinuingTimeLimit` actually. Exceeding the number of timesteps should not be a terminal signal.
